cucumber-core/src/main/java/io/cucumber/core/plugin/RerunFormatter.java

```diff
@@ -1,65 +1,48 @@
 package io.cucumber.core.plugin;
 
 import io.cucumber.core.feature.FeatureWithLines;
+import io.cucumber.messages.types.Envelope;
+import io.cucumber.messages.types.TestCaseFinished;
+import io.cucumber.messages.types.TestStepResult;
+import io.cucumber.messages.types.TestStepResultStatus;
 import io.cucumber.plugin.ConcurrentEventListener;
 import io.cucumber.plugin.event.EventPublisher;
-import io.cucumber.plugin.event.TestCase;
-import io.cucumber.plugin.event.TestCaseFinished;
-import io.cucumber.plugin.event.TestRunFinished;
+import io.cucumber.query.Query;
 
 import java.io.File;
 import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.util.ArrayList;
-import java.util.Collection;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.Map;
+import java.util.Set;
 
 import static io.cucumber.core.feature.FeatureWithLines.create;
+import static java.util.Objects.requireNonNull;
 
 /**
  * Formatter for reporting all failed test cases and print their locations
  * Failed means: results that make the exit code non-zero.
  */
 public final class RerunFormatter implements ConcurrentEventListener {
 
-    private final UTF8PrintWriter out;
-    private final Map<URI, Collection<Integer>> featureAndFailedLinesMapping = new LinkedHashMap<>();
+    private final PrintWriter writer;
+    private final Map<String, Set<Integer>> featureAndFailedLinesMapping = new LinkedHashMap<>();
+    private final Query query = new Query();
```

Comment:
Using query here seems over kill but as pickles don't have their location build in, this is the easiest way to get it. Consider putting the pickle location in the message.